funtools/man/man1/funimage.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "funimage 1"
.TH funimage 1 "April 14, 2011" "version 1.4.5" "SAORD Documentation"
.SH "NAME"
funimage \- create a FITS image from a Funtools data file
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBfunimage\fR [\-a] <iname> <oname> [bitpix=n]
\&\fBfunimage\fR [\-l] <iname> <oname> <xcol:xdims> <ycol:ydims> <vcol> [bitpix=n]
\&\fBfunimage\fR [\-p x|y] <iname> <oname> [bitpix=n]
.SH "OPTIONS"
.IX Header "OPTIONS"
.Vb 3
\&  \-a       # append to existing output file as an image extension
\&  \-l       # input is a list file containing xcol, ycol, value
\&  \-p [x|y] # project along x or y axis to create a 1D image
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBfunimage\fR creates a primary \s-1FITS\s0 image from the specified
\&\s-1FITS\s0 Extension
and/or
Image Section
of a \s-1FITS\s0 file, or from an
Image Section
of a non-FITS array, or from a raw event file.
.PP
The first argument to the program specifies the \s-1FITS\s0 input image,
array, or raw event file to process.  If \*(L"stdin\*(R" is specified, data are
read from the standard input. Use Funtools
Bracket Notation to specify \s-1FITS\s0 extensions, image sections, and
filters.  The second argument is the output \s-1FITS\s0 file.  If \*(L"stdout\*(R" is
specified, the \s-1FITS\s0 image is written to the standard output.  By
default, the output pixel values are of the same data type as those of the
input file (or type \*(L"int\*(R" when binning a table), but this can be
overridden using an optional third argument of the form:
.PP
.Vb 1
\&  bitpix=n
.Ve
.PP
where n is 8,16,32,\-32,\-64, for unsigned char, short, int, float and double,
respectively.
.PP
If the input data are of type image, the appropriate section is
extracted and blocked (based on how the 
Image Section is specified), and
the result is written to the \s-1FITS\s0 primary image.  When an integer
image containing the \s-1BSCALE\s0 and \s-1BZERO\s0 keywords is converted to float,
the pixel values are scaled and the scaling keywords are deleted from the
output header. When converting integer scaled data to integer
(possibly of a different size), the pixels are not scaled and the
scaling keywords are retained.
.PP
If the input data is a binary table or raw event file, these are
binned into an image, from which a section is extracted and blocked,
and written to a primary \s-1FITS\s0 image.  In this case, it is necessary to
specify the two columns that will be used in the 2D binning.  This can
be done on the command line using the \fBbincols=(x,y)\fR keyword:
.PP
.Vb 1
\&  funcnts "foo.ev[EVENTS,bincols=(detx,dety)]"
.Ve
.PP
The full form of the \fBbincols=\fR specifier is:
.PP
.Vb 1
\&  bincols=([xname[:tlmin[:tlmax:[binsiz]]]],[yname[:tlmin[:tlmax[:binsiz]]]])
.Ve
.PP
where the tlmin, tlmax, and binsiz specifiers determine the image binning
dimensions:
.PP
.Vb 2
\&  dim = (tlmax - tlmin)/binsiz     (floating point data)
\&  dim = (tlmax - tlmin)/binsiz + 1 (integer data)
.Ve
.PP
Using this syntax, it is possible to bin any two columns of a binary
table at any bin size.  Note that the tlmin, tlmax, and binsiz
specifiers can be omitted if \s-1TLMIN\s0, \s-1TLMAX\s0, and \s-1TDBIN\s0 header parameters
(respectively) are present in the \s-1FITS\s0 binary table header for the
column in question. Note also that if only one parameter is specified,
it is assumed to be tlmax, and tlmin defaults to 1. If two parameters
are specified, they are assumed to be tlmin and tlmax.
See Binning \s-1FITS\s0 Binary Tables and Non-FITS
Event Files for more information about binning parameters.
.PP
By default, a new 2D \s-1FITS\s0 image file is created and the image is written
to the primary \s-1HDU\s0.  If the \fB\-a\fR (append) switch is specified,
the image is appended to an existing \s-1FITS\s0 file as an \s-1IMAGE\s0 extension.
(If the output file does not exist, the switch is effectively ignored
and the image is written to the primary \s-1HDU\s0.)  This can be useful in a
shell programming environment when processing multiple \s-1FITS\s0 images
that you want to combine into a single final \s-1FITS\s0 file.
.PP
\&\fBfunimage\fR also can take input from a table containing columns of
x, y, and value (e.g., the output from \fBfundisp \-l\fR which
displays each image x and y and the number of counts at that
position.) When the \fB\-l\fR (list) switch is used, the input file is
taken to be a \s-1FITS\s0 or \s-1ASCII\s0 table containing (at least) three columns
that specify the x and y image coordinates and the value of that
image pixel. In this case, \fBfunimage\fR requires four extra
arguments: xcolumn:xdims, ycolumn:ydims, vcolumn and bitpix=n. The x
and y col:dim information takes the form: 
.PP
.Vb 3
\&  name:dim               # values range from 1 to dim
\&  name:min:max           # values range from min to max
\&  name:min:max:binsiz    # dimensions scaled by binsize
.Ve
.PP
In particular, the min value should be used whenever the
minimum coordinate value is something other than one. For example:
.PP
.Vb 1
\&  funimage \-l foo.lst foo.fits xcol:0:512 ycol:0:512 value bitpix=-32
.Ve
.PP
The list feature also can be used to read unnamed columns from standard
input: simply replace the column name with a null string. Note
that the dimension information is still required:
.PP
.Vb 5
\&  funimage \-l stdin foo.fits "":0:512 "":0:512 "" bitpix=-32
\&  240 250 1
\&  255 256 2
\&  ...
\&  ^D
.Ve
.PP
The list feature provides a simple way to generate a blank image.
If you pass a Column-based Text File

ChangeLog
2016-01-12 waj

	* upgraded TEA to 3.9
	* numerous compiler fixes for 64bit

2008-11-14  Jeff Hobbs  <jeffh@ActiveState.com>

	**** 2.10 TAGGED FOR RELEASE ****

	* generic/tkTable.c: Set NO_XSETCLIP universally as Tk 8.5/Xft
	seems to ignore XSetClipRectangles.  This may slow things down a
	little, but draws correctly. [Bug 1805350]

	* UPGRADING.txt (removed): no longer need this old v1 upgrade doc.

	* README.txt: remove refs to mac classic stuff, update notes

	* configure, configure.in, tclconfig/tcl.m4: Update to TEA 3.7
	* tests/tkTable.test, unix/tktable.spec, generic/version.h:
	* demos/loadtable.tcl: Update version to 2.10
	* Makefile.in: update dist target, include library/tktable.py

	* demos/tktable.py: note that this version is outdated
	* library/tktable.py (NEW): Added new version of tktable wrapper
	for Python/Tkinter from GPolo.  Not 100% compatible with previous
	version. [Patch 2244167]

	* generic/tkTableTag.c (Table_TagCmd): remove interp->result set
	work-around for old bug as i->result is no longer exposed in 8.6.

	* generic/tkTable.c:     Use safe Tcl_SetObjResult instead of trying
	* generic/tkTableCell.c: to set existing objResult value.
	* generic/tkTableCmds.c: 

2006-07-10  Jeff Hobbs  <jeffh@ActiveState.com>

	* license.txt (bourbon_ware): Be less restrictive in the
	bourbon_ware clause.

2006-01-25  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure.in, configure, tclconfig/tcl.m4: update to TEA 3.5

2006-01-22  Jeff Hobbs  <jeffh@ActiveState.com>

	* tclconfig/tcl.m4, configure: update to TEA 3.4 interim

2005-12-13  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/tkTableCmds.c (Table_ActivateCmd): remove potential
	undefined behavior warning.

2005-12-02  Jeff Hobbs  <jeffh@ActiveState.com>

	* tclconfig/tcl.m4, configure.in, configure: update to TEA 3.4

2005-05-11  Jeff Hobbs  <jeffh@ActiveState.com>

	* Makefile.in (html): switch to groff to fix botched html output.
	* doc/tkTable.html: regen [Bug 1198478]
	* doc/tkTable.n: use '' around \\ items.

2005-03-18  Jeff Hobbs  <jeffh@ActiveState.com>

	* Makefile.in (AR): use @AR@, improve html doc target
	* configure, configure.in, tclconfig/tcl.m4: TEA 3.2 patch update

	* doc/tkTable.n, doc/kTable.html: doc cleanup

2005-02-01  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure, configure.in, tclconfig/tcl.m4: update to TEA 3.2

2004-11-18  Jeff Hobbs  <jeffh@ActiveState.com>

	* library/tkTable.tcl (::tk::table::SelectAll): remove invalid ref
	to HandleType proc. [Bug 1068874] (ade)

2004-07-22  Jeff Hobbs  <jeffh@ActiveState.com>

	**** 2.9 TAGGED FOR RELEASE ****

	* configure, tclconfig/tcl.m4: update tcl.m4 TEA_PATH_X that adds
	Aqua build recognition and TEA_WINDOWINGSYSTEM variable.

2004-07-20  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/version.h, unix/tktable.spec:    updated for v2.9
	* demos/loadtable.tcl, tests/tkTable.test:

	* Makefile.in, configure.in, configure, tclconfig/tcl.m4: update
	to TEA 3.1, use CFLAGS instead of EXTRA_CFLAGS and bump the
	version number of Tktable to 2.9.

	* generic/tkTable.h: add #include <ctype.h>

	* generic/tkTableCell.c (TableGetCellValue, TableSetCellValue): 
	Do faster cache checks when DATA_CACHE is the only data source.
	In fallback from command that errors to array var, retrieve array
	value immediately on current get|set request.
	When setting the value with cache and array sources, don't do the
	caching branch as the array trace will do that already.
	(TableMoveCellValue): Do the efficient cache value management when
	cache is the only data source, otherwise the regular fallback has
	all the necessary checks. [Bug #919872]

	* generic/tkTable.c (TableVarProc): allow setting NULL value in
	cache hash from array trace.

2004-06-10  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/tkTableWin.c (EmbWinDisplay): make min window displayed
	have width/height 1 pixel min (was 3px)

	* generic/tkTable.c (TableDisplay): Don't draw when we don't have
	the space for it. [Bug 747038]

	* doc/tkTable.html, doc/tkTable.n: improve docs to note
	requirement of specifying a data source. [Bug 963204]

2004-02-17  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/tkTableCmds.c (Table_ActivateCmd): test implementation
	of an "unactivate" ($table activate "") command to remove the
	active cell. (becroft)

2004-02-13  Jeff Hobbs  <jeffh@ActiveState.com>

	* library/tkTable.tcl (::tk::table::Button1): account for what
	resizeborders are set when selecting cell [Bug 876320] (ferenc) 
	(<Prior> <Next>): activate topleft, not @0,0 (accounts for title

funtools/funtools-1.4.7/funimage.c
/*
 *	Copyright (c) 1999-2005 Smithsonian Astrophysical Observatory
 */

#include <funtoolsP.h>

extern char *optarg;
extern int optind;

#define FORCE_IMAGE -1

/* list definitions */
#define X_DEFAULT          "X"
#define Y_DEFAULT          "Y"
#define VAL_DEFAULT        "VAL"
#define COL1_DEFAULT       "COL1"
#define COL2_DEFAULT       "COL2"
#define COL3_DEFAULT       "COL3"

/* list event structure */
typedef struct evstruct{
  double x, y, val;
} *Ev, EvRec;

/* list structure */
typedef struct List {
  Fun fun;
  char *colname[3];
  double tlmin[2];
  double tlmax[2];
  double binsiz[2]; 
  int dim[2];
  double tscale[2]; 
  double tzero[2];
  int scaled[2];
} List;

#ifdef ANSI_FUNC
static void
doimproj(int dim, int bitpix, void *buf, int i, int j, void *pbuf, int idx)
#else
static void
doimproj(dim, bitpix, buf, i, j, pbuf, idx)
     int dim;
     int bitpix;
     void *buf;
     int i;
     int j;
     void *pbuf;
     int idx;
#endif
{
  switch(bitpix/FT_WORDLEN){
  case TY_CHAR:
    ((unsigned char *)pbuf)[idx] += ((unsigned char *)buf)[(j*dim)+i];
    break;
  case TY_USHORT:
    ((unsigned short *)pbuf)[idx] += ((unsigned short *)buf)[(j*dim)+i];
    break;
  case TY_SHORT:
    ((short *)pbuf)[idx] += ((short *)buf)[(j*dim)+i];
    break;
  case TY_INT:
    ((int *)pbuf)[idx] += ((int *)buf)[(j*dim)+i];
    break;
  case TY_LONG:
#if HAVE_LONG_LONG == 0
    gerror(stderr, 
	   "64-bit data support not built (long long not available)\n");
#endif
    ((longlong *)pbuf)[idx] += ((longlong *)buf)[(j*dim)+i];
    break;
  case TY_FLOAT:
    ((float *)pbuf)[idx] += ((float *)buf)[(j*dim)+i];
    break;
  case TY_DOUBLE:
    ((double *)pbuf)[idx] += ((double *)buf)[(j*dim)+i];
    break;
  default:
    break;
  }
}

#ifdef ANSI_FUNC
static void
usage (char *fname)
#else
static void usage(fname)
     char *fname;
#endif
{
  fprintf(stderr, "usage:\n");
  fprintf(stderr,
   "%s [-a] iname oname [bitpix=n[,bscale=false]]\n",
	  fname);
  fprintf(stderr,
  "%s -l [-a] iname oname xcol:xdims ycol:ydims vcol bitpix=n\n",
	  fname);
  fprintf(stderr, "optional switches:\n");
  fprintf(stderr, "  -a       # append to existing output file as an image extension\n");
  fprintf(stderr, "  -l       # input is a list file containing xcol, ycol, value\n");
  fprintf(stderr, "  -p [x|y] # project along x or y axis to create a 1D image\n");
  fprintf(stderr, "\n");
  fprintf(stderr, "Specify output image data type using bitpix=n, where n is:\n");
  fprintf(stderr, "   8      # unsigned char\n");
  fprintf(stderr, "  16      # short\n");
  fprintf(stderr, "  32      # int\n");
  fprintf(stderr, " -32      # float\n");
  fprintf(stderr, " -64      # double\n");
  fprintf(stderr, "(Defaults: same as input image, or 32 (int) for tables.)\n");
  fprintf(stderr, "\n");
  fprintf(stderr, "List format requires x, y column information of the form:\n");
  fprintf(stderr, "  name:dim               # values range from 1 to dim\n");
  fprintf(stderr, "  name:min:max           # values range from min to max\n");
  fprintf(stderr, "  name:min:max:binsiz    # dimensions scaled by binsize\n");
  fprintf(stderr, "\n(version: %s)\n", FUN_VERSION);
  exit(1);
}

#ifdef ANSI_FUNC
static List *
ColumnInfo(Fun fun, char *colstr[])
#else
static List *
ColumnInfo(fun, colstr)
     Fun fun;
     char *colstr[];
#endif
{
  int i;
  int ip;
  char tbuf[SZ_LINE];
  List *list;

  /* allocate list struct */
  if( !(list = (List *)xcalloc(sizeof(List), 1)) ){
    gerror(stderr, "can't allocate list record\n");
    return NULL;
  }

  /* process column info */
  newdtable(":");
  for(i=0; i<3; i++){
    ip = 0;
    /* column name */
    if( !*colstr[i] || (*colstr[i] == ':') || !word(colstr[i], tbuf, &ip) ){
      switch(i){
      case 0:
	if(FunColumnLookup(fun, COL1_DEFAULT, 0, 
			   NULL, NULL, NULL, NULL, NULL, NULL))
	  strcpy(tbuf, COL1_DEFAULT);
	else{
	  strcpy(tbuf, X_DEFAULT);
	}
	break;
      case 1:
	if(FunColumnLookup(fun, COL2_DEFAULT, 0, 
			   NULL, NULL, NULL, NULL, NULL, NULL))
	  strcpy(tbuf, COL2_DEFAULT);
	else{
	  strcpy(tbuf, Y_DEFAULT);
	}
	break;
      case 2:
	if(FunColumnLookup(fun, COL3_DEFAULT, 0, 
			   NULL, NULL, NULL, NULL, NULL, NULL))
	  strcpy(tbuf, COL3_DEFAULT);
	else{
	  strcpy(tbuf, VAL_DEFAULT);
	}
	break;
      }
    }
    list->colname[i] = xstrdup(tbuf);

    /* back up to the previous ':' as required by column dimension routine */
    if( ip && (lastdelim() == ':') ) ip--;
    /* column dimensions */
    if( colstr[i][ip] != '\0' ){
      switch(i){
      case 0:
      case 1:
	_FunColumnDims(&colstr[i][ip], 'I', 
		       &list->tlmin[i], &list->tlmax[i],
		       &list->binsiz[i], &list->dim[i],
		       &list->tscale[i], &list->tzero[i], &list->scaled[i]);
	if( list->tlmin[i] && list->tlmax[i] )	
	  list->dim[i] = (int)tldim(list->tlmin[i], list->tlmax[i],
				    list->binsiz[i], 'I');
	else
	  list->dim[i] = 0;
	break;
      case 2:
	break;
      }
    }
    /* must have dimesion info for x and y */
    else{
      switch(i){
      case 0:
      case 1:
	gerror(stderr, "dimension specification missing for %s\n", colstr[i]);
	return NULL;
      case 2:
	break;
      }
    }
  }
  freedtable();

  /* set up to read the specified columns */
  FunColumnSelect(fun, sizeof(EvRec), NULL,
		  list->colname[0],  "D", "r", FUN_OFFSET(Ev, x),
		  list->colname[1],  "D", "r", FUN_OFFSET(Ev, y),
		  list->colname[2],  "D", "r", FUN_OFFSET(Ev, val),
		  NULL);

  /* return list */
  return list;
}

#ifdef ANSI_FUNC
int 
main (int argc, char **argv)
#else
int
main(argc, argv)
     int argc;
     char **argv;
#endif
{
  int c;
  int i, j;
  int args;
  int dim1, dim2;
  int idx;
  int dim=0;
  int maxidx=0;
  int type=0;
  int bitpix=0;
  int doscale=1;
  int doproj=0;
  char *s, *t;
  char *mode=NULL;
  char *tmode=NULL;
  char *omode="w";
  char *iname=NULL;
  char *oname=NULL;
  char *colstr[3];
  char *buf=NULL;
  char *pbuf=NULL;
  char tbuf[SZ_LINE];
  char newname[SZ_LINE];
  Fun fun, fun2;

  /* list variables */
  int dolist=0;
  int ix, iy;
  int got;
  List *list=NULL;
  EvRec evrec;

  /* exit on gio errors */
  if( !getenv("GERROR")  )
    setgerror(2);

  /* we want the args in the same order in which they arrived, and
     gnu getopt sometimes changes things without this */
  putenv("POSIXLY_CORRECT=true");

  /* process switch arguments */
  while ((c = getopt(argc, argv, "ailp:")) != -1){
    switch(c){
    case 'a':
      omode = "a";
    case 'i':
      type = FORCE_IMAGE;
      break;
    case 'l':
      dolist = 1;
      break;
    case 'p':
      doproj = *optarg;
      break;
    }
  }

  /* sanity checks */
  if( doproj ){
    if( dolist ) gerror(stderr, "-p and -l are mutually exclusive\n");
    switch(doproj){
    case 'x':
    case 'y':
      break;
    case 'X':
    case 'Y':
      doproj = tolower(doproj);
      break;
    default:
      gerror(stderr, "proj specifier must be 'x' or 'y'\n");
      break;
    }
  }

  /* process arguments */
  args = argc - optind;
  if( args < 2 ) usage(argv[0]);
  iname = argv[optind+0];
  oname = argv[optind+1];
  if( dolist ){
    if( args !=  6 ) usage(argv[0]);
    colstr[0] = argv[optind+2];
    colstr[1] = argv[optind+3];
    colstr[2] = argv[optind+4];
    mode = xstrdup(argv[optind+5]);
  }
  else{
    if( args > 2 ) mode = xstrdup(argv[optind+2]);
  }

  /* check for $NEW and rewrite name as /dev/null[TEXT(),...] */
  if( !strncmp(iname, "$NEW[", 5 ) || !strncmp(iname, "$MASK[", 6 )  ){
    /* for masks, we add to the mode */
    if( !strncmp(iname, "$MASK[", 6 )  ){
      if( mode && *mode ){
	snprintf(tbuf, SZ_LINE-1, "%s,mask=all", mode);
	xfree(mode);
	mode = xstrdup(tbuf);
      }
      else{
	mode = xstrdup("mask=all,bitpix=8");
      }
    }
    /* re-write file name as an empty text table */
    memset(newname, 0, SZ_LINE);
    strncpy(newname, "/dev/null[TEXT(x:", SZ_LINE-1);
    s = strchr(iname, '[')+1;
    t = strchr(s, ',');
    strncat(newname, s, t-s);
    strncat(newname, ",y:", SZ_LINE-strlen(newname));
    if (t != NULL) {
      s = ++t;
      t = strchr(s, ',');
      strncat(newname, s, t-s);
    }
    strncat(newname, "),", SZ_LINE-strlen(newname));
    if (t != NULL) {
	s = ++t;
	strncat(newname, s, SZ_LINE-strlen(newname));
    }
    iname = newname;
  }

  /* open the input FITS file */
  if( !(fun = FunOpen(iname, "rc", NULL)) ){
    gerror(stderr, "can't FunOpen input file (or find extension): %s\n",
	   iname);
  }

  /* open the output FITS image, preparing to copy input params */
  if( !(fun2 = FunOpen(oname, omode, fun)) ){
    gerror(stderr, "can't FunOpen output file: %s\n", oname);
  }

  /* list: read x, y, val from list file */
  if( dolist ){
    /* make sure bitpix=n is specified */
    tmode = xstrdup(mode);
    if( _FunKeyword(tmode, "bitpix", NULL, tbuf, SZ_LINE) ){
      bitpix = atoi(tbuf);
    }
    else{
      gerror(stderr, "missing bitpix specification: '%s'\n", mode);
    }
    if( tmode ) xfree(tmode);

    /* parse column string */
    if( !(list = ColumnInfo(fun, colstr)) ){
      gerror(stderr, "can't get column info\n");
    }

    /* allocate space for the image */
    if( list->dim[1] )
      maxidx = list->dim[0]*list->dim[1];
    else
      maxidx = list->dim[0];
    if( !(buf=(char *)xcalloc(maxidx, ft_sizeof(bitpix))) ){
      gerror(stderr, "can't allocate image buffer of size %d\n", maxidx);
    }

    /* process events */
    while( FunTableRowGet(fun, &evrec, 1, NULL, &got) && got){
      /* get x, y indices */
      ix = (int)((evrec.x - list->tlmin[0]) / list->binsiz[0]);
      if( list->dim[1] )
	iy = (int)((evrec.y - list->tlmin[1]) / list->binsiz[1]);
      else
	iy = 0;
      /* make sure this pixel is within the image */
      idx = iy*list->dim[0]+ix;
      if( (idx < 0) || (idx >= maxidx) ){
	gerror(stderr, 
	"indices (%d,%d) outside of image array (%d,%d) [tlmin problem?]\n",
	       ix, iy, list->dim[0], list->dim[1]);
      }
      /* add pixel value */
      switch(bitpix){
      case 8:
	((unsigned char *)buf)[idx] += (unsigned char)evrec.val;
	break;
      case 16:
	((short *)buf)[idx] += (short)evrec.val;
	break;
      case 32:
	((int *)buf)[idx] += (int)evrec.val;
	break;
      case -32:
	((float *)buf)[idx] += (float)evrec.val;
	break;
      case -64:
	((double *)buf)[idx] += evrec.val;
	break;
      }
    }
    /* write the output image, updating the FITS header from the orig file */
    if( !FunImagePut(fun2, buf, list->dim[0], list->dim[1], bitpix, NULL) ){
	gerror(stderr, "could not FunImagePut: %s\n", argv[2]);
    }
  }
  /* no list: bin input data into an image */
  else{
    /* get required information from funtools structure */
    FunInfoGet(fun, 
	       FUN_SECT_BITPIX, &bitpix,
	       FUN_SECT_DIM1, &dim1, FUN_SECT_DIM2, &dim2, 0);
    if( type != FORCE_IMAGE ) FunInfoGet(fun, FUN_TYPE, &type, 0);
    /* set dim2 to 1 for a 1D image */
    if( dim2 == 0 ){
      /* oops, can't project a 1D image */
      if( doproj )
	gerror(stderr, "can't project a 1D image\n");
      else
	dim2 = 1;
    }
    /* dimension of projected 1D image */
    if( doproj ){
      switch(doproj ){
      case 'x':
	dim = dim1;
	break;
      case 'y':
	dim = dim2;
	break;
      }
      /* allocate buffer for 1D projected image */
      pbuf = (char *)xcalloc(dim, ABS(bitpix)/FT_WORDLEN);
    }

    /* process specific type of data */
    switch(type){
    case FUN_IMAGE:
    case FUN_ARRAY:
      /* if input data is int */
      if( bitpix >= 0 ){
	/* start by assuming we don't want to apply bscale ... */
	doscale = 0;
	if( mode && *mode ){
	  tmode = xstrdup(mode);
	  /* but if output data is float, we will apply bscale */
	  if( _FunKeyword(tmode, "bitpix", NULL, tbuf, SZ_LINE) ){
	    if( (i=atoi(tbuf)) < 0 ){
	      /* apple bscale */
	      doscale=1;
	    }
	  }
	  if( tmode ) xfree(tmode);
	}
      }
      /* for float data, we apply bscale (won't actually happen) */
      else{
	doscale = 1;
      }
      if( !doscale ){
	if( mode && *mode ){
	  mode = xrealloc(mode, strlen(mode)+SZ_LINE);
	  strcat(mode, ",bscale=false");
	}
	else{
	  mode = xstrdup("bscale=false");
	}
      }
      switch(doproj ){
      case 0:
	buf = NULL;
	for(j=1; j<=dim2; j++){
	  /* extract and bin 1 row of the data section into an image buffer */
	  if( !(buf = FunImageRowGet(fun, buf, j, j, mode)) )
	    gerror(stderr, "can't FunImageRowGet: %s\n", iname);
	  /* write output image, updating the FITS header from the orig file */
	  if( !FunImageRowPut(fun2, buf, j, j, 0, 0, 0, NULL) )
	    gerror(stderr, "can't FunImageRowPut: %s\n", oname);
	}
	break;
      case 'x':
      case 'y':
	buf = NULL;
	for(j=0; j<dim2; j++){
	  /* extract and bin 1 row of the data section into an image buffer */
	  if( !(buf = FunImageRowGet(fun, buf, j+1, j+1, mode)) )
	    gerror(stderr, "can't FunImageRowGet: %s\n", iname);
	  /* project each pixel of each line */
	  for(i=0; i<dim1; i++){
	    idx = (doproj == 'x') ? i : j;
	    doimproj(dim, bitpix, (void *)buf, i, 0, (void *)pbuf, idx);
	  }
	}
	/* write output image, updating the FITS header from the orig file */
	if( !FunImagePut(fun2, pbuf, dim, 0, bitpix, NULL) )
	  gerror(stderr, "can't FunImagePut: %s\n", oname);
	break;
      }
      break;
      
    case FUN_TABLE:
    case FUN_EVENTS:
    case FORCE_IMAGE:
      switch(doproj ){
      case 0:
	/* extract and bin the data section into an image buffer */
	if( !(buf = FunImageGet(fun, NULL, mode)) )
	  gerror(stderr, "can't FunImageGet: %s\n", iname);
	/* write output image, updating the FITS header from the orig file */
	if( !FunImagePut(fun2, buf, 0, 0, 0, NULL) )
	  gerror(stderr, "can't FunImagePut: %s\n", oname);
	break;
      case 'x':
      case 'y':
	/* extract and bin the data section into an image buffer */
	if( !(buf = FunImageGet(fun, NULL, mode)) )
	  gerror(stderr, "can't FunImageGet: %s\n", iname);
	/* project each pixel of each line */
	for(j=0; j<dim2; j++){
	  for(i=0; i<dim1; i++){
	    idx = (doproj == 'x') ? i : j;
	    doimproj(dim, bitpix, (void *)buf, i, j, (void *)pbuf, idx);
	  }
	}
	/* write output image, updating the FITS header from the orig file */
	if( !FunImagePut(fun2, pbuf, dim, 0, bitpix, NULL) )
	  gerror(stderr, "can't FunImagePut: %s\n", oname);
	break;
      }
      break;

    default:
      gerror(stderr, "internal error: unknown type\n");
      break;
      }
  }

  /* we can explicitly flush remaining extensions that need copying ... */
  FunFlush(fun2, "copy=remaining");

  /* ... since we already flushed, we can close files in any order */
  FunClose(fun);
  FunClose(fun2);

  /* free up space */
  if( dolist ){
    /* free up memory */
    for(i=0; i<3; i++){
      if( list->colname[i] ) xfree(list->colname[i]);
    }
    xfree(list);
  }
  if( buf )  xfree(buf);
  if( pbuf ) xfree(pbuf);
  if( mode ) xfree(mode);

  return(0);
}
